<a href="https://colab.research.google.com/github/alexsasik/for-dz/blob/main/online_sex_work.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Исследовательская задача**

Датасет, выбранный для исследования, представляет собой большое количество данных с форума для взрослых. Однако сами создатели написали в своей работе, что основная исследовательская задача не выполнена. Только небольшому проценту исследуемым был рассчитан риск их вовлечения в оказание сексуальных услуг онлайн. 

Следовательно, нашей основной задачей становится выявление факторов, влияющих на риск, на основе уже исследованных людей и расчет риска для неисследованных. Кроме того, датасет был с технической ошибкой, вследствие которой после собранных данных шло огромное количество пустых мест. Поэтому для начала необходимо убрать их, и начать работать уже с чистым датасетом.

## **Переменные:**

*   **User_ID** - идентификационный номер пользователя на форуме
*   **Gender** - пол пользователя, male/female
*   **Age** - возраст пользователя в годах
*   **Location** - Местоположение пользователя, в целях конфиденциальности показано как название точки (A, B и т.д.)
*   **Verification** - является пользователь верифицированнным на форуме или нет
*   **Sexual_orientation** - сексуальная ориентация пользователя (гетеро-, би- или гомосексуальная)
*   **Sexual_polarity** - сексуальная полярность пользователя
*   **Looking_for** - предпочтения пользователя в выборе пары
*   **Points_Rank** - количество очков на форуме
*   **Last_login** - время последнего пребывания в сети с момента обращения к статистике форума
*   **Member_since** - дата в формате ДД.ММ.ГГГГ, в которую пользователь зарегистрировался на форуме
*   **Number_of_comments_in_public_forum** - число комментариев, оставленных пользователем
*   **Time_spent_chatting_H:M** - время, проведенное за общением на форуме в формате Часы:Минуты
*   **Number_of_adbertisments_posted** - количество размещенных рекламных сообщений
*   **Number_of_offline_meetings_attended** - количество посещенных оффлайн-свиданий после общения на форуме
*   **Profile_pictures** - количество фотографий в профиле пользователя
*   **Friends_ID_list** - список идентификационных номеров друзей пользователя с разделителем в виде запятой
*   **Risk** - риск быть вовлеченным в предоставление сексуальных услуг онлайн











## **Очистка датасета от пустых значений**

In [20]:
import pandas as pd

In [21]:
data = pd.read_csv('https://github.com/alexsasik/for-dz/raw/main/online_sex_work.csv')
data

/usr/local/lib/python3.7/dist-packages/IPython/core/interactiveshell.py:3326: DtypeWarning: Columns (1,2,3,4,5,6,7,8,9,10,11,12,16,17) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


User_ID Gender   Age Location  Verification Sexual_orientation  \
0        10386.0   male  34,6        A  Non_Verified         Homosexual   
1           14.0   male  32,2        J  Non_Verified       Heterosexual   
2        16721.0   male  33,6        K  Non_Verified       Heterosexual   
3        16957.0   male    34        H  Non_Verified       Heterosexual   
4        17125.0   male  39,5        B  Non_Verified       Heterosexual   
...          ...    ...   ...      ...           ...                ...   
1048570      NaN    NaN   NaN      NaN           NaN                NaN   
1048571      NaN    NaN   NaN      NaN           NaN                NaN   
1048572      NaN    NaN   NaN      NaN           NaN                NaN   
1048573      NaN    NaN   NaN      NaN           NaN                NaN   
1048574      NaN    NaN   NaN      NaN           NaN                NaN   

        Sexual_polarity Looking_for Points_Rank      Last_login Member_since  \
0                Switch         Men          50  before_10_days    17.9.2012   
1              Dominant       Women         518   before_1_days    1.11.2009   
2              Dominant       Women         150   before_3_days     1.4.2013   
3              Dominant       Women         114   before_4_days     8.4.2013   
4              Dominant       Women         497   before_5_days    14.4.2013   
...                 ...         ...         ...             ...          ...   
1048570             NaN         NaN         NaN             NaN          NaN   
1048571             NaN         NaN         NaN             NaN          NaN   
1048572             NaN         NaN         NaN             NaN          NaN   
1048573             NaN         NaN         NaN             NaN          NaN   
1048574             NaN         NaN         NaN             NaN          NaN   

        Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
0                                       32                    0:2   
1                                      710                   3:45   
2                                       25                   2:15   
3                                      107                 359:22   
4                                      600                   0:21   
...                                    ...                    ...   
1048570                                NaN                    NaN   
1048571                                NaN                    NaN   
1048572                                NaN                    NaN   
1048573                                NaN                    NaN   
1048574                                NaN                    NaN   

         Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
0                                   0.0                                  0.0   
1                                   9.0                                  0.0   
2                                   1.0                                  1.0   
3                                   1.0                                  0.0   
4                                   0.0                                  6.0   
...                                 ...                                  ...   
1048570                             NaN                                  NaN   
1048571                             NaN                                  NaN   
1048572                             NaN                                  NaN   
1048573                             NaN                                  NaN   
1048574                             NaN                                  NaN   

         Profile_pictures                                    Friends_ID_list  \
0                     0.0                                              18260   
1                     0.0          11778,32024,43768,23969,27318,45884,31277   
2                    45.0                                  19805,21721,19802   
3                     1.0  40847,3818

Необходимо убрать все пустые строки. Такими будут те, где не значится **User_ID**. Отсутствие ID означает отсутствие пользователя.

In [22]:
firstnan = data[data['User_ID'].isnull()].iloc[0]
firstnan

User_ID                                NaN
Gender                                 NaN
Age                                    NaN
Location                               NaN
Verification                           NaN
Sexual_orientation                     NaN
Sexual_polarity                        NaN
Looking_for                            NaN
Points_Rank                            NaN
Last_login                             NaN
Member_since                           NaN
Number_of_Comments_in_public_forum     NaN
Time_spent_chating_H:M                 NaN
Number_of_advertisments_posted         NaN
Number_of_offline_meetings_attended    NaN
Profile_pictures                       NaN
Friends_ID_list                        NaN
Risk                                   NaN
Name: 28831, dtype: object

Создадим **срез** датасета, который будет очищен от пустых значений и назовем его **cdata** (clean data). С этим срезом мы будем работать дальше, чтобы пустые значения не искажали статистику.

In [23]:
cdata = data.iloc[:28831]
cdata

User_ID  Gender   Age Location  Verification Sexual_orientation  \
0      10386.0    male  34,6        A  Non_Verified         Homosexual   
1         14.0    male  32,2        J  Non_Verified       Heterosexual   
2      16721.0    male  33,6        K  Non_Verified       Heterosexual   
3      16957.0    male    34        H  Non_Verified       Heterosexual   
4      17125.0    male  39,5        B  Non_Verified       Heterosexual   
...        ...     ...   ...      ...           ...                ...   
28826   9962.0    male  27,2        J  Non_Verified       Heterosexual   
28827   9964.0    male  46,4        J  Non_Verified          bicurious   
28828   9966.0    male  28,8        C  Non_Verified       Heterosexual   
28829   9968.0    male  31,5        J  Non_Verified           bisexual   
28830    998.0  female  38,7        F  Non_Verified       Heterosexual   

      Sexual_polarity    Looking_for Points_Rank       Last_login  \
0              Switch            Men          50   before_10_days   
1            Dominant          Women         518    before_1_days   
2            Dominant          Women         150    before_3_days   
3            Dominant          Women         114    before_4_days   
4            Dominant          Women         497    before_5_days   
...               ...            ...         ...              ...   
28826          Switch          Women           0   before_23_days   
28827       Submisive  Men_and_Women          15  before_597_days   
28828       Submisive          Women          15    before_4_days   
28829       Submisive            Men          30  before_665_days   
28830        Dominant         Nobody          20  before_157_days   

      Member_since Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
0        17.9.2012                                 32                    0:2   
1        1.11.2009                                710                   3:45   
2         1.4.2013                                 25                   2:15   
3         8.4.2013                                107                 359:22   
4        14.4.2013                                600                   0:21   
...            ...                                ...                    ...   
28826     5.9.2012                                  0                    0:3   
28827     5.9.2012                                  0                    0:0   
28828     5.9.2012                                  0                    0:0   
28829     1.9.2012                                  0                   2:54   
28830    20.5.2010                                  0                    0:9   

       Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
0                                 0.0                                  0.0   
1                                 9.0                                  0.0   
2                                 1.0                                  1.0   
3                                 1.0                                  0.0   
4                                 0.0                                  6.0   
...                               ...                                  ...   
28826                             1.0                                  0.0   
28827                             0.0                                  0.0   
28828                             0.0                                  0.0   
28829                             4.0                                  0.0   
28830                             0.0                                  0.0   

       Profile_pictures                                    Friends_ID_list  \
0                   0.0                                              18260   
1                   0.0          11778,32024,43768,23969,27318,45884,31277   
2                  45.0                                  19805,21721,19802   
3                   1.0  40847,38183,9507,42259,5807,28118,24848,37170,...   
4

## **Фильтры**

Поскольку нас интересуют демографические и ориентационные данные пользователей, зарегистрированных на взрослом онлайн-форуме, нам следует сделать несколько фильтров по этим критериям. Это нужно для того, чтобы понимать, что и как может влиять на риск быть вовлеченным в предоставление или получение сексуальных услуг онлайн.

Начнем с пользователей с гомосексуальной ориентацией.

In [24]:
cdata[cdata.Sexual_orientation == 'Homosexual']

User_ID Gender   Age Location  Verification Sexual_orientation  \
0      10386.0   male  34,6        A  Non_Verified         Homosexual   
55      9479.0   male  31,8        A  Non_Verified         Homosexual   
83      5648.0   male  41,4        K  Non_Verified         Homosexual   
152    10090.0   male  33,6        A  Non_Verified         Homosexual   
167    10114.0   male  59,1        A  Non_Verified         Homosexual   
...        ...    ...   ...      ...           ...                ...   
28720   9806.0   male  48,8        A  Non_Verified         Homosexual   
28751   9858.0   male  31,7        H  Non_Verified         Homosexual   
28757   9864.0   male  34,6        M  Non_Verified         Homosexual   
28788   9899.0   male  23,8        H  Non_Verified         Homosexual   
28789   9900.0   male  43,9        A  Non_Verified         Homosexual   

      Sexual_polarity Looking_for Points_Rank        Last_login Member_since  \
0              Switch         Men          50    before_10_days    17.9.2012   
55             Switch         Men         268    before_64_days    25.8.2012   
83          Submisive         Men          62     before_4_days    24.6.2012   
152         Submisive      Nobody           0  before_1683_days     8.9.2012   
167          Dominant      Nobody           0   before_220_days     9.9.2012   
...               ...         ...         ...               ...          ...   
28720        Dominant      Nobody          15   before_193_days     2.9.2012   
28751       Submisive         Men           0  before_1699_days     3.9.2012   
28757        Dominant      Nobody           0  before_1433_days     3.9.2012   
28788       Submisive      Nobody           0   before_800_days     4.9.2012   
28789        Dominant         Men           0  before_1101_days     4.9.2012   

      Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
0                                     32                    0:2   
55                                    92                    0:5   
83                                    30                    0:0   
152                                    0                    0:0   
167                                    0                    0:0   
...                                  ...                    ...   
28720                                  0                    0:0   
28751                                  0                    0:0   
28757                                  0                    0:0   
28788                                  0                    0:0   
28789                                  0                    0:0   

       Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
0                                 0.0                                  0.0   
55                                0.0                                  0.0   
83                                0.0                                  0.0   
152                               0.0                                  0.0   
167                               0.0                                  0.0   
...                               ...                                  ...   
28720                             0.0                                  0.0   
28751                             0.0                                  0.0   
28757                             0.0                                  0.0   
28788                             1.0                                  0.0   
28789                             0.0                                  0.0   

       Profile_pictures                                    Friends_ID_list  \
0                   0.0                                              18260   
55                 33.0  11476,21928,5073,9660,42259,14635,21430,6094,7...   
83                  7.0                             12766,9479,17701,26991   
152                 0.0                                                NaN   
167                 0.0              

Теперь посмотрим, как они взаимодействуют с форумом и собеседниками с него. Отфильтруем неверифицированных пользователей с гомосексуальной ориентацией.

In [25]:
cdata[(cdata.Sexual_orientation == 'Homosexual') & (cdata.Verification == 'Verified')]

User_ID  Gender   Age Location Verification Sexual_orientation  \
7745   21430.0    male  36,4        A     Verified         Homosexual   
8151   22013.0    male  36,8        L     Verified         Homosexual   
11387  26991.0    male    34        A     Verified         Homosexual   
12733  29175.0    male    30        A     Verified         Homosexual   
15971  34066.0    male  31,9        C     Verified         Homosexual   
18986   3845.0  female  26,9        K     Verified         Homosexual   
22566  43819.0    male  34,4        L     Verified         Homosexual   
23872  45782.0    male  28,8        A     Verified         Homosexual   
25268  47801.0    male  33,6        M     Verified         Homosexual   

      Sexual_polarity       Looking_for Points_Rank       Last_login  \
7745        Submisive            Nobody         166    before_3_days   
8151        Submisive               Men         178  before_224_days   
11387       Submisive               Men         155    before_4_days   
12733       Submisive            Nobody          50  before_315_days   
15971       Submisive               Men         266    before_4_days   
18986        Dominant             Women         397    before_3_days   
22566          Switch  Nobody_but_maybe         165   before_84_days   
23872       Submisive            Nobody         150   before_25_days   
25268       Submisive               Men         195    before_3_days   

      Member_since Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
7745     24.9.2013                                 37                   3:38   
8151    18.10.2013                                  3                    0:0   
11387    19.5.2014                                  0                    0:0   
12733    21.8.2014                                  0                    0:0   
15971    17.3.2015                                  2                    1:8   
18986    1.11.2011                                164                    0:0   
22566     3.9.2016                                  0                    0:0   
23872   14.12.2016                                  0                    0:0   
25268    23.3.2017                                  0                    0:0   

       Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
7745                              0.0                                  2.0   
8151                              1.0                                  0.0   
11387                             0.0                                  1.0   
12733                             0.0                                  0.0   
15971                             1.0                                  0.0   
18986                             1.0                                  0.0   
22566                             1.0                                  1.0   
23872                             0.0                                  0.0   
25268                             0.0                                  0.0   

       Profile_pictures                                    Friends_ID_list  \
7745                9.0  21357,11989,5245,17220,9479,25684,9436,4808,29...   
8151                8.0                                              20563   
11387               1.0              10510,9479,21430,9436,4808,29342,5648   
12733               0.0                                                NaN   
15971              22.0                                  39142,36469,47774   
18986              42.0                                                NaN   
22566               3.0                                  20966,41481,26333   
23872               0.0                                                NaN   
25268               8.0                                                NaN   

               Risk  
7745   unknown_risk  
8151   unknown_risk  
11387  unknown_risk  
12733  unknown_risk  
15971  unknown_risk  
18986  unknown_risk  
22566  unknown_risk  
23872  unknown_r

Далее посмотрим, кто из них встречался хотя бы один раз с кем-то из собеседников в реальной жизни.

In [26]:
cdata[(cdata.Sexual_orientation == 'Homosexual') & (cdata.Number_of_offline_meetings_attended > 0) & (cdata.Verification == 'Verified')]

User_ID Gender   Age Location Verification Sexual_orientation  \
7745   21430.0   male  36,4        A     Verified         Homosexual   
11387  26991.0   male    34        A     Verified         Homosexual   
22566  43819.0   male  34,4        L     Verified         Homosexual   

      Sexual_polarity       Looking_for Points_Rank      Last_login  \
7745        Submisive            Nobody         166   before_3_days   
11387       Submisive               Men         155   before_4_days   
22566          Switch  Nobody_but_maybe         165  before_84_days   

      Member_since Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
7745     24.9.2013                                 37                   3:38   
11387    19.5.2014                                  0                    0:0   
22566     3.9.2016                                  0                    0:0   

       Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
7745                              0.0                                  2.0   
11387                             0.0                                  1.0   
22566                             1.0                                  1.0   

       Profile_pictures                                    Friends_ID_list  \
7745                9.0  21357,11989,5245,17220,9479,25684,9436,4808,29...   
11387               1.0              10510,9479,21430,9436,4808,29342,5648   
22566               3.0                                  20966,41481,26333   

               Risk  
7745   unknown_risk  
11387  unknown_risk  
22566  unknown_risk

Выберем тех пользователей с нетрадиционной сексуальной ориентацией, у которых высокий риск. Это еще риск, определенный авторами датасета, так что данные могут быть неполными.

In [27]:
cdata[(cdata.Sexual_orientation == 'Homosexual') & (cdata.Risk == 'High_risk')]

User_ID Gender   Age Location  Verification Sexual_orientation  \
83   5648.0   male  41,4        K  Non_Verified         Homosexual   

   Sexual_polarity Looking_for Points_Rank     Last_login Member_since  \
83       Submisive         Men          62  before_4_days    24.6.2012   

   Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
83                                 30                    0:0   

    Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
83                             0.0                                  0.0   

    Profile_pictures         Friends_ID_list       Risk  
83               7.0  12766,9479,17701,26991  High_risk

Отфильтруем пользователей в поиске и мужчин, и женщин. Посмотрим, сколько будет людей с бисексуальной ориентацией, у которых в будущем мы можем выявить высокий риск.

In [28]:
cdata[(cdata.Looking_for == 'Men_and_Women')]

User_ID  Gender   Age Location  Verification Sexual_orientation  \
35     43605.0  female  24,9        H  Non_Verified           bisexual   
51      6732.0    male    36        S  Non_Verified          bicurious   
56      9550.0  female  49,9        B      Verified           bisexual   
64     21180.0  female  33,8        S  Non_Verified          bicurious   
66     23282.0    male  27,3        A  Non_Verified          bicurious   
...        ...     ...   ...      ...           ...                ...   
28736   9835.0    male  33,1        A  Non_Verified           bisexual   
28762   9870.0    male    24        O  Non_Verified           bisexual   
28798   9919.0    male  31,5        S  Non_Verified          bicurious   
28822   9951.0    male  32,4        A  Non_Verified           bisexual   
28827   9964.0    male  46,4        J  Non_Verified          bicurious   

      Sexual_polarity    Looking_for Points_Rank        Last_login  \
35             Switch  Men_and_Women          78     before_3_days   
51          Submisive  Men_and_Women           3     before_7_days   
56           Dominant  Men_and_Women         429    before_89_days   
64             Switch  Men_and_Women          12  before_1306_days   
66          Submisive  Men_and_Women          28     before_4_days   
...               ...            ...         ...               ...   
28736       Submisive  Men_and_Women           0  before_1686_days   
28762       Submisive  Men_and_Women          15    before_20_days   
28798       Submisive  Men_and_Women          15  before_1630_days   
28822        Dominant  Men_and_Women           0  before_1515_days   
28827       Submisive  Men_and_Women          15   before_597_days   

      Member_since Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
35       23.8.2016                                 62                  105:4   
51        9.7.2012                                  4                   0:37   
56       27.8.2012                                267                  58:57   
64       13.9.2013                                  1                  41:48   
66      11.12.2013                                 13                    0:0   
...            ...                                ...                    ...   
28736     2.9.2012                                  0                    0:0   
28762     3.9.2012                                  0                    0:0   
28798     4.9.2012                                  0                    0:0   
28822     5.9.2012                                  0                    6:6   
28827     5.9.2012                                  0                    0:0   

       Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
35                                0.0                                  0.0   
51                                3.0                                  0.0   
56                                0.0                                  0.0   
64                                0.0                                  0.0   
66                                1.0                                  0.0   
...                               ...                                  ...   
28736                             0.0                                  0.0   
28762                             0.0                                  0.0   
28798                             0.0                                  0.0   
28822                             0.0                                  0.0   
28827                             0.0                                  0.0   

       Profile_pictures                                    Friends_ID_list  \
35                 10.0                            44736,18996,27733,42022   
51                  0.0                      11079,12617,29822,10270,43268   
56                  5.0  2460,6829,23361,10510,9436,6077,20566,8746,963...   
64                  0.0                                  21224,1660

## **Сортировка**

Также мы можем отсортировать датасет для того, чтобы увидеть, влияет ли более высокая активность на форуме или предельные демографические признаки (самый молодой или самый старый) на риски.

За активность возьмем количество размещенных рекламных сообщений.

In [29]:
cdata.sort_values(by=['Number_of_advertisments_posted'], ascending=False)

User_ID  Gender   Age Location  Verification Sexual_orientation  \
8231    2213.0    male  37,5        A  Non_Verified       Heterosexual   
9031   23314.0    male  32,3        J      Verified       Heterosexual   
3625   15385.0    male  52,9        A  Non_Verified       Heterosexual   
6       1726.0    male  36,7        B  Non_Verified          bicurious   
2208    1325.0    male    50        A  Non_Verified           bisexual   
...        ...     ...   ...      ...           ...                ...   
10155  25115.0    male  32,3        G  Non_Verified       Heterosexual   
10154  25112.0    male  44,5        L  Non_Verified       Heterosexual   
10153  25111.0    male  41,6        K  Non_Verified          bicurious   
10152   2511.0    male  57,3        L  Non_Verified       Heterosexual   
28830    998.0  female  38,7        F  Non_Verified       Heterosexual   

      Sexual_polarity Looking_for Points_Rank        Last_login Member_since  \
8231         Dominant       Women          50     before_4_days    14.1.2011   
9031           Switch       Women          82    before_11_days   12.12.2013   
3625        Submisive       Women          15     before_3_days    17.2.2013   
6           Submisive       Women          18     before_1_days   16.10.2010   
2208         Dominant      Nobody           0  before_1107_days     5.8.2010   
...               ...         ...         ...               ...          ...   
10155          Switch      Nobody           0  before_1156_days    18.2.2014   
10154          Switch      Nobody           0  before_1135_days    18.2.2014   
10153       Submisive      Nobody           0  before_1166_days    18.2.2014   
10152        Dominant      Nobody          15     before_1_days    13.3.2011   
28830        Dominant      Nobody          20   before_157_days    20.5.2010   

      Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
8231                                   0                    0:0   
9031                                   0                    9:5   
3625                                   0                    0:0   
6                                      1                 122:20   
2208                                   0                   0:20   
...                                  ...                    ...   
10155                                  0                    0:0   
10154                                  0                    0:0   
10153                                  0                    0:0   
10152                                  0                    0:0   
28830                                  0                    0:9   

       Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
8231                             21.0                                  0.0   
9031                             18.0                                  0.0   
3625                             18.0                                  0.0   
6                                17.0                                  0.0   
2208                             16.0                                  0.0   
...                               ...                                  ...   
10155                             0.0                                  0.0   
10154                             0.0                                  0.0   
10153                             0.0                                  0.0   
10152                             0.0                                  0.0   
28830                             0.0                                  0.0   

       Profile_pictures          Friends_ID_list          Risk  
8231                0.0                      NaN  unknown_risk  
9031                1.0  13408,23365,18827,22311  unknown_risk  
3625                0.0                      NaN  unknown_risk  
6                   0.0                      NaN       No_risk  
2208                0.0                      NaN  unknown_risk  
...                 ...  

Можем посмотреть и социальную активность, то есть отсортировать по наибольшему количеству посещенных встреч в реальной жизни.

In [30]:
cdata.sort_values(by=['Number_of_offline_meetings_attended'], ascending=False)

User_ID  Gender   Age Location  Verification Sexual_orientation  \
12097  28195.0    male  38,9        L  Non_Verified       Heterosexual   
9496   24073.0    male  37,9        G      Verified       Heterosexual   
27034   6848.0    male  36,8        K      Verified       Heterosexual   
6077   18996.0    male  25,9        O      Verified       Heterosexual   
9688   24389.0  female  27,1        G  Non_Verified           bisexual   
...        ...     ...   ...      ...           ...                ...   
9680   24374.0    male  37,1        G  Non_Verified       Heterosexual   
9679   24373.0    male  29,3        O  Non_Verified       Heterosexual   
9678    2437.0    male  45,9        G  Non_Verified       Heterosexual   
9677   24368.0    male  52,4        H  Non_Verified       Heterosexual   
28830    998.0  female  38,7        F  Non_Verified       Heterosexual   

      Sexual_polarity       Looking_for Points_Rank        Last_login  \
12097        Dominant             Women         167     before_5_days   
9496         Dominant             Women         210     before_8_days   
27034        Dominant  Nobody_but_maybe         315     before_1_days   
6077           Switch            Nobody         883     before_1_days   
9688           Switch             Women         150    before_34_days   
...               ...               ...         ...               ...   
9680        Submisive            Nobody           0  before_1197_days   
9679           Switch            Nobody           0  before_1190_days   
9678         Dominant            Nobody           0  before_2139_days   
9677         Dominant            Nobody           0  before_1197_days   
28830        Dominant            Nobody          20   before_157_days   

      Member_since Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
12097    10.7.2014                                183               2 642:51   
9496      7.1.2014                                  1                  28:53   
27034    10.7.2012                                179               1 137:33   
6077     20.6.2013                                226               1 417:34   
9688     18.1.2014                                  1                 265:40   
...            ...                                ...                    ...   
9680     18.1.2014                                  0                    0:0   
9679     18.1.2014                                  0                    0:0   
9678     28.2.2011                                  0                    0:0   
9677     17.1.2014                                  0                    0:0   
28830    20.5.2010                                  0                    0:9   

       Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
12097                             1.0                                 23.0   
9496                              1.0                                 16.0   
27034                             0.0                                 15.0   
6077                              1.0                                 13.0   
9688                              1.0                                 11.0   
...                               ...                                  ...   
9680                              0.0                                  0.0   
9679                              0.0                                  0.0   
9678                              0.0                                  0.0   
9677                              0.0                                  0.0   
28830                             0.0                                  0.0   

       Profile_pictures                                    Friends_ID_list  \
12097               0.0  38183,23365,41786,20197,33755,43133,42730,1973...   
9496                2.0  42407,33755,43144,18069,16604,31776,21404,2081...   
27034               6.0  32759,24922,41946,46745,31540,20843,43133,2593...   
6077                2.0  21357,

Как мы видим, риск пока неизвестен, так что, вероятно, предельные значения в активности не влияют на него резко позитивно или негативно. 
Попробуем посмотреть по возрасту:

In [31]:
cdata.sort_values(by=['Age']).head()

User_ID  Gender   Age Location  Verification Sexual_orientation  \
25179  47677.0    male  18,1        A  Non_Verified       Heterosexual   
25671  48431.0  female  18,1        L  Non_Verified          bicurious   
25712  48496.0    male  18,2        C  Non_Verified          bicurious   
25603  48329.0  female  18,2        K  Non_Verified          bicurious   
25614  48342.0    male  18,2        A  Non_Verified          bicurious   

      Sexual_polarity Looking_for Points_Rank      Last_login Member_since  \
25179       Submisive      Nobody           0  before_41_days    18.3.2017   
25671       Submisive      Nobody           0   before_9_days    20.4.2017   
25712          Switch      Nobody           0   before_6_days    23.4.2017   
25603       Submisive         Men          15   before_1_days    16.4.2017   
25614        Dominant      Nobody           0  before_12_days    17.4.2017   

      Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
25179                                  0                    0:0   
25671                                  0                    0:0   
25712                                  0                    0:0   
25603                                  0                   0:10   
25614                                  0                   0:13   

       Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
25179                             0.0                                  0.0   
25671                             0.0                                  0.0   
25712                             0.0                                  0.0   
25603                             0.0                                  0.0   
25614                             0.0                                  0.0   

       Profile_pictures          Friends_ID_list          Risk  
25179               0.0                      NaN  unknown_risk  
25671               0.0                      NaN  unknown_risk  
25712               0.0                      NaN  unknown_risk  
25603               0.0  47827,48284,21437,46447  unknown_risk  
25614               0.0                      NaN  unknown_risk

К сожалению, риски неизвестны и по самым молодым пользователям, которым, как можно судить по возрасту, чуть больше 18 лет. 
Раз сортировка по предельным значениям не помогла, то посмотрим частотные данные, которые дадут нам представление как минимум о той аудитории, которая находится на данном форуме.

## **Частотные таблицы**

Создадим частотные таблицы для всех переменных, содержащих категориальные данные, кроме переменной **Risk**, так как риск по ней не рассчитан. Для переменных **Age, Points_Rank, Member_since, Number_of_Comments_in_public_forum, Time_spent_chating_H:M, Friends_ID_list**	количество значений будет слишком большое, из-за чего данные будет сложно интерпретировать. Из них данные можно будет извлечь в новые объединенные переменные (например, разбить возраст на промежутки в 10 лет). В то же время такие переменные, как **Last_login, Number_of_advertisments_posted, Number_of_offline_meetings_attended, Profile_pictures** не обязательно будут иметь слишком большой разброс значений, поэтому, для проверки этой гипотезы, мы создадим для них частотные таблицы.

Эти таблицы дадут нам понимание о преобладающих демографических показателях пользователей форума и их преобладающих сексуальных предпочтениях.

In [32]:
cdata['Gender'].value_counts(dropna=False)

male      22382
female     6445
NaN           4
Name: Gender, dtype: int64

Эта таблица показывает, что большая потребность в форуме для взрослых встречается у пользователей мужского пола, поскольку их почти в 3,5 раза больше, чем женщин.

In [33]:
cdata['Location'].value_counts(dropna=False)

A                     10944
B                      3572
G                      2211
L                      2210
M                      1261
K                      1209
C                      1176
H                      1167
E                       951
O                       892
J                       849
F                       822
N                       623
S                       484
D                       455
A kraj                    2
Jihomoravsk_ý kraj        1
NaN                       1
South Moravian            1
Name: Location, dtype: int64

Частотная таблица по локации пользователей показывает, что целевой аудиторией форума является один населенный пункт/область, так как почти треть пользователей находятся в локации А (в чистом датасете 28830 пользователей).

In [34]:
cdata['Verification'].value_counts(dropna=False)

Non_Verified     28228
Verified           602
Administrator        1
Name: Verification, dtype: int64

Из данной таблицы можно заключить, что подтвержденных пользователей довольно мало по сравнению с неподтвержденными. Кроме того, мы видим, что на форуме есть только 1 администратор.

In [35]:
cdata['Sexual_orientation'].value_counts(dropna=False)

Heterosexual    20441
bisexual         4391
bicurious        2992
Homosexual       1006
NaN                 1
Name: Sexual_orientation, dtype: int64

Среди пользователей форума примерно две трети являются гетеросексуальными, а их соотношение с гомосексуальными примерно равно 20:1.

In [36]:
cdata['Sexual_polarity'].value_counts(dropna=False)

Submisive    10878
Dominant     10474
Switch        7478
NaN              1
Name: Sexual_polarity, dtype: int64

Количество пользователей, проявляющих только доминантную или только подчиняющуюся сексуальную активность, примерно равно и составляет чуть больше двух третей, тогда как оставшуюся неполную треть занимают пользователи, у которых проявление сексуальной активности непостоянно.

In [37]:
cdata['Looking_for'].value_counts(dropna=False)

Nobody              19928
Women                4896
Nobody_but_maybe     1512
Men_and_Women        1074
Men                   996
NaN                   425
Name: Looking_for, dtype: int64

Из частотной таблицы можно заключить, что большая часть пользователей находится на форуме скорее не для знакомств, а для общения или чтения дискуссий. Наиболее же часто встречающийся пол искомого партнера - женский, что предсказуемо, так как большую часть аудитории форума составляют гетеросексуальные мужчины.

In [38]:
cdata['Last_login'].value_counts(dropna=False)

before_1_days       1392
before_3_days        961
before_4_days        314
before_6_days        237
before_5_days        233
                    ... 
before_2113_days       1
before_2376_days       1
before_2097_days       1
before_2026_days       1
before_2398_days       1
Name: Last_login, Length: 2360, dtype: int64

Среди топ-5 дат последнего онлайна все даты вписываются в промежуток недели до момента создания датасета. В то же время количество пользователей, которые были в сети хотя бы раз за неделю до сбора данных, составляет примерно 9-10% (мы пропускаем тех, кто был за 2 и за 7 дней до сбора данных в сети, так как это не очень важно). Это дает нам понимание, что примерно такой процент от всей аудитории форума достаточно активен.

In [39]:
cdata['Number_of_advertisments_posted'].value_counts(dropna=False)

0.0     25164
1.0      2984
2.0       453
3.0       114
4.0        48
5.0        28
6.0        12
8.0         6
7.0         5
10.0        3
9.0         3
11.0        3
14.0        2
18.0        2
16.0        1
15.0        1
17.0        1
21.0        1
Name: Number_of_advertisments_posted, dtype: int64

Подавляющее большинство пользователей форума ни разу не публиковало рекламного сообщения, а количество тех, кто публиковал хотя бы одно, меньше примерно в 9-10 раз. Дважды публиковало сообщения уже в 5 раз меньшее количество, чем единожды. 

In [40]:
cdata['Number_of_offline_meetings_attended'].value_counts(dropna=False)

0.0     28302
1.0       293
2.0        91
3.0        50
4.0        34
5.0        21
6.0        11
7.0         9
9.0         6
10.0        4
11.0        3
8.0         3
13.0        1
16.0        1
23.0        1
15.0        1
Name: Number_of_offline_meetings_attended, dtype: int64

По этой частотной таблице можно сказать, что пользователи в основном пользуются форумом для онлайн-общения. Более 98% ни разу не встречались со своими собеседниками в реальной жизни.

In [41]:
cdata['Profile_pictures'].value_counts(dropna=False)

0.0     26351
1.0       936
2.0       396
3.0       257
4.0       185
        ...  
67.0        1
53.0        1
45.0        1
34.0        1
40.0        1
Name: Profile_pictures, Length: 63, dtype: int64

По полученным данным можно судить о том, что пользователи форума для взрослых не любят публиковать свои личные данные. 91,5% пользователей не устанавливал фотографию профиля и сохранял визуальную анонимность.

## **Новая переменная**

Поскольку мы хотим высчитать вероятности риска, то нам найти какие-то тенденции среди тех, у кого высокий риск, уже указанный создателями датасета. Для начала получим срез тех, у кого высокий риск.

In [55]:
high_risk = cdata['Risk'].isin(['High_risk'])
cdata[high_risk]

User_ID  Gender   Age Location   Verification Sexual_orientation  \
57  12262.0    male  35,7        A   Non_Verified       Heterosexual   
58  12944.0    male  41,1        A   Non_Verified       Heterosexual   
59   1353.0    male  49,6        A   Non_Verified       Heterosexual   
60  13767.0    male  27,9        A   Non_Verified          bicurious   
61  15172.0    male  52,3        E   Non_Verified       Heterosexual   
62  17871.0    male  27,3        F   Non_Verified       Heterosexual   
63  19635.0  female    45        A   Non_Verified       Heterosexual   
64  21180.0  female  33,8        S   Non_Verified          bicurious   
65  22592.0    male  35,2        B   Non_Verified       Heterosexual   
66  23282.0    male  27,3        A   Non_Verified          bicurious   
67  23362.0    male  24,9        B       Verified       Heterosexual   
68   2460.0    male  47,2        A   Non_Verified       Heterosexual   
69  26123.0    male  42,4        F   Non_Verified           bisexual   
70  34414.0    male  34,3        A   Non_Verified       Heterosexual   
71   3474.0    male  36,3        M       Verified       Heterosexual   
72  34922.0    male    30        B       Verified       Heterosexual   
73  35359.0    male  36,1        A   Non_Verified       Heterosexual   
74  39099.0    male  34,5        J   Non_Verified       Heterosexual   
75  39386.0    male  46,8        F   Non_Verified       Heterosexual   
76  42730.0  female  23,3        L       Verified       Heterosexual   
77  43155.0  female  41,6        K   Non_Verified       Heterosexual   
78  43606.0    male  28,6        A   Non_Verified          bicurious   
79  43811.0    male  20,8        A   Non_Verified           bisexual   
80  44987.0    male  37,7        A   Non_Verified       Heterosexual   
81  47131.0    male  28,3        M   Non_Verified          bicurious   
82  48218.0    male  43,4        H   Non_Verified       Heterosexual   
83   5648.0    male  41,4        K   Non_Verified         Homosexual   
84   5675.0    male  36,2        N   Non_Verified       Heterosexual   
85   5782.0    male  29,9        A   Non_Verified       Heterosexual   
86   6035.0    male  52,3        A   Non_Verified       Heterosexual   
87   6288.0  female  28,8        A   Non_Verified          bicurious   
88   9383.0  female  34,2        L       Verified       Heterosexual   
89      1.0    male  32,4        J  Administrator                NaN   

   Sexual_polarity       Looking_for Points_Rank        Last_login  \
57       Submisive             Women           0     before_4_days   
58       Submisive  Nobody_but_maybe           0     before_1_days   
59       Submisive            Nobody          15    before_11_days   
60       Submisive            Nobody           5     before_6_days   
61       Submisive            Nobody           0   before_424_days   
62       Submisive             Women          15     before_1_days   
63        Dominant            Nobody          25    before_11_days   
64          Switch     Men_and_Women          12  before_1306_days   
65       Submisive             Women           0  before_1255_days   
66       Submisive     Men_and_Women          28     before_4_days   
67       Submisive             Women         512     before_4_days   
68       Submisive            Nobody         184     before_3_days   
69       Submisive  Nobody_but_maybe          15     before_1_days   
70       Submisive             Women          25     before_1_days   
71       Submisive  Nobody_but_maybe         307     before_2_days   
72       Submisive  Nobody_but_maybe         471     before_6_days   
73       Submisive             Women          48     before_3_days   
74          Switch             Women          50     before_1_days   
75          Switch             Women          55     before_3_days   
76        Dominant  Nobody_but_maybe         318     before_1_days   
77        Dominant  Nobody_but_maybe           5    before_32_days   
78       Submis

Далее мы можем сделать частотные таблицы по некоторым из столбцов, в которых прослеживается явное преобладание одного значения, чтобы, выявив преобладающие значения, определить степень риска у людей, которые подходят по двум или нескольким этим значениям.

In [56]:
cdata[high_risk]['Gender'].value_counts(dropna=False)

male      27
female     6
Name: Gender, dtype: int64

In [57]:
cdata[high_risk]['Sexual_orientation'].value_counts(dropna=False)

Heterosexual    23
bicurious        6
bisexual         2
Homosexual       1
NaN              1
Name: Sexual_orientation, dtype: int64

In [58]:
cdata[high_risk]['Sexual_polarity'].value_counts(dropna=False)

Submisive    23
Switch        5
Dominant      4
NaN           1
Name: Sexual_polarity, dtype: int64

In [59]:
cdata[high_risk]['Looking_for'].value_counts(dropna=False)

Women               13
Nobody              11
Nobody_but_maybe     6
Men_and_Women        2
Men                  1
Name: Looking_for, dtype: int64

In [60]:
cdata[high_risk]['Last_login'].value_counts(dropna=False)

before_1_days       12
before_3_days        7
before_4_days        4
before_11_days       2
before_6_days        2
before_424_days      1
before_1306_days     1
before_1255_days     1
before_2_days        1
before_32_days       1
before_9_days        1
Name: Last_login, dtype: int64

На основе этих данных можно построить **гипотезу**, что, скорее всего, имеется риск у тех, кто является гетеросексуальным мужчиной с подчиняющейся сексуальной полярностью, который либо ищет женщину, либо не ищет никого, и который заходил в последний раз не более 4-х дней назад. Пока что мы используем наиболее упрощенный вариант, который будет показывать высокий риск в случае совпадения 3 или более факторов и средний риск при совпадении 2 факторов. Учитывая, что мы учитываем пока что не все факторы из таблицы, мы не будем ставить отсутствие риска всем, кто не подходит под упрощенную модель.

Создадим функцию, которая поможет нам определить риски.

In [63]:
def risk(gender, SO, SP, Lf, Ll):
  summa = 0
  new_Ll = int("".join(c for c in Ll if c.isdecimal()))
  if gender == 'male':
    summa += 1
  if SO == 'Heterosexual':
    summa += 1
  if SP == 'Submisive':
    summa += 1
  if Lf == 'Women' or Lf == 'Nobody':
    summa += 1
  if new_Ll <= 4:
    summa += 1
  if summa >= 4:
    summa = 0
    return 'high_risk'
  elif summa == 3:
    summa = 0
    return 'higher_potential_risk'
  elif summa == 2:
    summa = 0
    return 'lower_potential_risk'
  else:
    summa = 0
    return 'cannot define yet'
  


Теперь применим функцию к нашему датасету и вызовем его, чтобы проверить, как работает наша методика расчета.

In [64]:
cdata['New_Risk'] = cdata[['Gender', 'Sexual_orientation', 'Sexual_polarity', 'Looking_for', 'Last_login']].apply(lambda cdata: risk(cdata['Gender'], cdata['Sexual_orientation'], cdata['Sexual_polarity'], cdata['Looking_for'], cdata['Last_login']), axis = 1)
cdata

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


User_ID  Gender   Age Location  Verification Sexual_orientation  \
0      10386.0    male  34,6        A  Non_Verified         Homosexual   
1         14.0    male  32,2        J  Non_Verified       Heterosexual   
2      16721.0    male  33,6        K  Non_Verified       Heterosexual   
3      16957.0    male    34        H  Non_Verified       Heterosexual   
4      17125.0    male  39,5        B  Non_Verified       Heterosexual   
...        ...     ...   ...      ...           ...                ...   
28826   9962.0    male  27,2        J  Non_Verified       Heterosexual   
28827   9964.0    male  46,4        J  Non_Verified          bicurious   
28828   9966.0    male  28,8        C  Non_Verified       Heterosexual   
28829   9968.0    male  31,5        J  Non_Verified           bisexual   
28830    998.0  female  38,7        F  Non_Verified       Heterosexual   

      Sexual_polarity    Looking_for Points_Rank       Last_login  \
0              Switch            Men          50   before_10_days   
1            Dominant          Women         518    before_1_days   
2            Dominant          Women         150    before_3_days   
3            Dominant          Women         114    before_4_days   
4            Dominant          Women         497    before_5_days   
...               ...            ...         ...              ...   
28826          Switch          Women           0   before_23_days   
28827       Submisive  Men_and_Women          15  before_597_days   
28828       Submisive          Women          15    before_4_days   
28829       Submisive            Men          30  before_665_days   
28830        Dominant         Nobody          20  before_157_days   

      Member_since Number_of_Comments_in_public_forum Time_spent_chating_H:M  \
0        17.9.2012                                 32                    0:2   
1        1.11.2009                                710                   3:45   
2         1.4.2013                                 25                   2:15   
3         8.4.2013                                107                 359:22   
4        14.4.2013                                600                   0:21   
...            ...                                ...                    ...   
28826     5.9.2012                                  0                    0:3   
28827     5.9.2012                                  0                    0:0   
28828     5.9.2012                                  0                    0:0   
28829     1.9.2012                                  0                   2:54   
28830    20.5.2010                                  0                    0:9   

       Number_of_advertisments_posted  Number_of_offline_meetings_attended  \
0                                 0.0                                  0.0   
1                                 9.0                                  0.0   
2                                 1.0                                  1.0   
3                                 1.0                                  0.0   
4                                 0.0                                  6.0   
...                               ...                                  ...   
28826                             1.0                                  0.0   
28827                             0.0                                  0.0   
28828                             0.0                                  0.0   
28829                             4.0                                  0.0   
28830                             0.0                                  0.0   

       Profile_pictures                                    Friends_ID_list  \
0                   0.0                                              18260   
1                   0.0          11778,32024,43768,23969,27318,45884,31277   
2                  45.0                                  19805,21721,19802   
3                   1.0  40847,38183,9507,42259,5807,28118,24848,37170,...   
4

И теперь, когда уже есть бета-версия рисков, можно посмотреть вырисовывающиеся частоты риска пользователей онлайн-форума оказаться вовлеченными в предоставление сексуальных услуг онлайн.

In [65]:
cdata['New_Risk'].value_counts(dropna=False)

higher_potential_risk    15228
lower_potential_risk      6157
high_risk                 5246
cannot define yet         2200
Name: New_Risk, dtype: int64

## **Меры центральной тенденции**

Рассчитаем меры центральной тенденции для основных переменных:

In [66]:
cdata.columns

Index(['User_ID', 'Gender', 'Age', 'Location', 'Verification',
       'Sexual_orientation', 'Sexual_polarity', 'Looking_for', 'Points_Rank',
       'Last_login', 'Member_since', 'Number_of_Comments_in_public_forum',
       'Time_spent_chating_H:M', 'Number_of_advertisments_posted',
       'Number_of_offline_meetings_attended', 'Profile_pictures',
       'Friends_ID_list', 'Risk', 'New_Risk'],
      dtype='object')

In [67]:
clmns = ['User_ID', 'Gender', 'Age', 'Location', 'Verification',
       'Sexual_orientation', 'Sexual_polarity', 'Looking_for', 'Points_Rank',
       'Last_login', 'Member_since', 'Number_of_Comments_in_public_forum',
       'Time_spent_chating_H:M', 'Number_of_advertisments_posted',
       'Number_of_offline_meetings_attended', 'Profile_pictures',
       'Friends_ID_list', 'Risk']

In [68]:
cdata['User_ID'].mode()

0      2035.0
1      2589.0
2      2773.0
3      3212.0
4      4349.0
       ...   
63    48026.0
64    48232.0
65    48326.0
66    48438.0
67    48490.0
Length: 68, dtype: float64